includes/payment-methods/class-wc-stripe-upe-payment-gateway.php

```diff
@@ -438,7 +438,9 @@ public function payment_fields() {
 			<?php
 			if ( $display_tokenization ) {
 				$this->tokenization_script();
-				$this->saved_payment_methods();
+				if ( $this->get_tokens() ) {
```

Comment:
This works well. My doubt is why, without this change, things work well on the regular checkout but not on the Pay for order page?

This seems to be handled on the client-side on the checkout page. It also works well in both the Checkout and Pay for order pages in WCPay and we don't check for `$this->get_tokens()` before calling `$this->saved_payment_methods()` there. Again, this works. I'm wondering if there's a small detail we're missing here for things to work as expected.

Reply:
During checkout the payment methods element gets re-rendered right after the page is loaded which ends up removing the options when they are not necessary. I could not find any JS that removes the options on the plugin apparently thats something that is done by WooCommerce itself. I'll take a deeper look to see if there's anything there.

I have written another solution with js only which would wait till everything loads and check if the options are necessary and hide it when not. But to me it makes more sense to not render it at all when not necessary

Reply:
I found that [this file](https://github.com/woocommerce/woocommerce/blob/release/5.9/assets/js/frontend/tokenization-form.js#L109) handles the form on checkout and it is triggered [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/classic/upe.js#L288)

So I copied that to Stripe plugin and it works too. Can you please take another look at this?


Resolved with 222fbdee73dfa62bc70b50d2b91628d19d251993